src/mlpack/methods/ann/ffn_impl.hpp

```diff
@@ -633,6 +634,77 @@ FFN<OutputLayerType, InitializationRuleType,
   return *this;
 };
 
+template<typename OutputLayerType, typename InitializationRuleType,
+         typename... CustomLayers>
+template<typename InputType, typename OutputType>
+void FFN<OutputLayerType, InitializationRuleType,
+    CustomLayers...>::Forward(
+    InputType&& input, OutputType&& output){
+  boost::apply_visitor(ForwardVisitor(std::move(arma::mat(input)), std::move(
+      boost::apply_visitor(outputParameterVisitor, network.front()))),
+      network.front());
+
+  for (size_t i = 1; i < network.size(); ++i)
+  {
+    boost::apply_visitor(ForwardVisitor(std::move(boost::apply_visitor(
+        outputParameterVisitor, network[i - 1])), std::move(
+        boost::apply_visitor(outputParameterVisitor, network[i]))), network[i]);
+  }
+
+  output = arma::mat(boost::apply_visitor(outputParameterVisitor, network.back()));
+}
+
+template<typename OutputLayerType, typename InitializationRuleType,
+         typename... CustomLayers>
+template<typename InputType, typename ErrorType, typename GradientType>
+void FFN<OutputLayerType, InitializationRuleType,
+    CustomLayers...>::Backward(
+  const InputType&& /* input */, ErrorType&& gy, GradientType&& g)
+{
+  boost::apply_visitor(BackwardVisitor(std::move(boost::apply_visitor(
+      outputParameterVisitor, network.back())), std::move(gy),
+      std::move(boost::apply_visitor(deltaVisitor, network.back()))),
+      network.back());
+
+  for (int i = network.size() - 2; i > 0; i--)
+  {
+    boost::apply_visitor(BackwardVisitor(std::move(boost::apply_visitor(
+        outputParameterVisitor, network[i])),
+        std::move(boost::apply_visitor(deltaVisitor, network[i + 1])),
+        std::move(boost::apply_visitor(deltaVisitor, network[i]))),
+        network[i]);
+  }
+
+  boost::apply_visitor(BackwardVisitor(std::move(boost::apply_visitor(
+      outputParameterVisitor, network.front())),
+      std::move(boost::apply_visitor(deltaVisitor, network[1])),
```

Comment:
Hey @zoq The failure is because of this line and the error is as follows 
```
test.out: /usr/include/boost/variant/detail/forced_return.hpp:47: T boost::detail::variant::forced_return() 
[with T = arma::Mat<double>&]: Assertion `false' failed.

```
Please If you can give me some insight about deltaVisitor how it work's it will be quite helpful for me.